File: LeetCode/JavaScript/B11-238-Product of Array Except Self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  const res = new Array(nums.length);
  let multi = 1;
  for (let i = 0; i < nums.length; i++) {
    res[i] = multi;
    multi *= nums[i];
  }
  multi = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    res[i] *= multi;
    multi *= nums[i];
  }

  return res;
};

console.log(productExceptSelf([1, 2, 3, 4]));

File: LeetCode/5회차/B11-238-Product of Array Except Self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        left, right = 0, len(nums) - 1
        left_multi, right_multi = 1, 1
        while right >= 0:
            res[left] *= left_multi
            res[right] *= right_multi
            left_multi *= nums[left]
            right_multi *= nums[right]
            left += 1
            right -= 1
        return res


print(Solution().productExceptSelf([1,2,3,4]))
